source/host/KTestSystem.cc

```diff
@@ -86,11 +89,100 @@ void TestDirector::parseSuite(EGG::RamStream &stream) {
 
         m_testCases.push(testCase);
     }
+
+    startNextTestCase();
+    m_sceneMgr->changeScene(0);
+}
+
+void KTestSystem::calc() {
+    m_sceneMgr->calc();
+}
+
+bool KTestSystem::run() {
+    bool success = true;
+
+    while (true) {
+        success &= runTest();
+
+        if (!popTestCase()) {
+            break;
+        }
+
+        // TODO: Use a system heap! We currently have a dependency on the scene heap
+        m_sceneMgr->destroyScene(m_sceneMgr->currentScene());
+        startNextTestCase();
+        m_sceneMgr->createScene(2, m_sceneMgr->currentScene());
+    }
+
+    return success;
+}
+
+void KTestSystem::parseOptions(int argc, char **argv) {
+    if (argc < 2) {
+        PANIC("Expected suite argument!");
+    }
+
+    for (int i = 0; i < argc; ++i) {
+        std::optional<Host::EOption> flag = Host::Option::CheckFlag(argv[i]);
+        if (!flag || *flag == Host::EOption::Invalid) {
+            WARN("Expected a flag! Got: %s", argv[i]);
+            continue;
+        }
+
+        switch (*flag) {
+        case Host::EOption::Suite: {
+            ASSERT(i + 1 < argc);
+
+            size_t size;
+            u8 *data = Abstract::File::Load(argv[++i], size);
+
+            if (size == 0) {
+                PANIC("Failed to load suite data!");
+            }
+
+            m_stream = EGG::RamStream(data, size);
+            m_stream.setEndian(std::endian::big);
+        } break;
+        case Host::EOption::Invalid:
+        default:
+            PANIC("Invalid flag!");
+            break;
+        }
+    }
+}
+
+KTestSystem *KTestSystem::CreateInstance() {
+    ASSERT(!s_instance);
+    s_instance = new KTestSystem;
+    return static_cast<KTestSystem *>(s_instance);
+}
+
+void KTestSystem::DestroyInstance() {
+    ASSERT(s_instance);
+    auto *instance = s_instance;
+    s_instance = nullptr;
+    delete instance;
+}
+
+KTestSystem *KTestSystem::Instance() {
+    return static_cast<KTestSystem *>(s_instance);
+}
+
+KTestSystem::KTestSystem() = default;
+
+KTestSystem::~KTestSystem() {
+    if (s_instance) {
+        s_instance = nullptr;
+        WARN("KTestSystem instance not explicitly handled!");
+    }
 }
 
-void TestDirector::init() {
+/// @brief Starts the next test case.
+void KTestSystem::startNextTestCase() {
+    constexpr u32 KRKG_SIGNATURE = 0x4b524b47; // KRKG
+
     size_t size;
-    u8 *krkg = Abstract::File::Load(testCase().krkgPath.data(), size);
+    u8 *krkg = Abstract::File::Load(getCurrentTestCase().krkgPath.data(), size);
```

Comment:
This is the only time we call `getCurrentTestCase()` with no active test case. An aliasing solution would be overengineering. In all other situations, this *does* return the current test case.

We're starting the next one because there is no current one, and the state is initialized with the previous one. If you'd like, I can add a detail to `getCurrentTestCase()` explaining that there is a situation where it returns the next test case if we're inactive.